nifi-nar-bundles/nifi-extension-utils/nifi-service-utils/src/main/java/org/apache/nifi/service/lookup/AbstractSingleAttributeBasedControllerServiceLookup.java

```diff
@@ -0,0 +1,162 @@
+ * <p>
+ * Lookup key is provided as a value in an attribute map (usually coming form a flowfile)
+ * with a predefined key (see {@link #lookupAttribute()}).
+ *
+ * @param <S> The type of service to be looked up
+ */
+public abstract class AbstractSingleAttributeBasedControllerServiceLookup<S extends ControllerService> extends AbstractControllerService {
+    protected volatile Map<String, S> serviceMap;
+
+    /**
+     * @return the Class that represents the type of service that will be returned by {@link #lookupService(Map)}
+     */
+    public abstract Class<S> getServiceType();
+
+    /**
+     * @return the name of attribute (usually from a flowfile) the value of which serves as the lookup key
+     * for the desired service (of type {@link S})
+     */
+    protected abstract String lookupAttribute();
+
+    /**
+     * Returns a ControllerService (of type {@link S}) based on the provided attributes map (usually by retrieving
+     *  a lookup attribute from it via {@link #lookupAttribute()} and use it to identify the appropriate service).
+     * @param attributes Map containing the lookup attribute based on which ControllerService is chosen
+     * @return the chosen ControllerService
+     */
+    public S lookupService(Map<String, String> attributes) {
+        if (attributes == null) {
+            throw new ProcessException("Attributes map is null");
+        } else if (!attributes.containsKey(lookupAttribute())) {
+            throw new ProcessException("Attributes must contain an attribute name '" + lookupAttribute() + "'");
+        }
+
+        Object lookupKey = Optional.of(lookupAttribute())
+                .map(attributes::get)
+                .orElseThrow(() -> new ProcessException(lookupAttribute() + " cannot be null or blank"));
+
+        S service = serviceMap.get(lookupKey);
+
+        if (service == null) {
+            throw new ProcessException("No " + getServiceName() + " found for " + lookupAttribute());
+        }
+
+        return service;
+    }
+
+    @OnEnabled
+    public void onEnabled(final ConfigurationContext context) {
+        Map<String, S> serviceMap = new HashMap<>();
+
+        context.getProperties().keySet().stream()
+                .filter(PropertyDescriptor::isDynamic)
+                .forEach(propertyDescriptor -> {
+                    S service = context.getProperty(propertyDescriptor).asControllerService(getServiceType());
+
+                    serviceMap.put(propertyDescriptor.getName(), service);
+                });
+
+        this.serviceMap = Collections.unmodifiableMap(serviceMap);
+    }
+
+    @OnDisabled
+    public void onDisabled() {
+        serviceMap = null;
+    }
+
+    @Override
+    protected PropertyDescriptor getSupportedDynamicPropertyDescriptor(final String propertyDescriptorName) {
+        return lookupKeyPropertyDescriptor(propertyDescriptorName);
+    }
+
+    @Override
+    protected Collection<ValidationResult> customValidate(ValidationContext context) {
```

Comment:
The lookup service does not check the status of the referenced controller services and it can be enabled when a service is in disabled state. Interestingly, the other way works: if I disable a referenced service, then the lookup service gets disabled.
It was not implemented in the original lookup services so it would rather be an improvement, maybe in a separate jira ticket.
It seems to me the framework only checks the controller services referenced via static properties.

Reply:
This a good catch! Lets make sure we file a JIRA to fix this at the framework level. I think the issue is that the dynamic properties aren't in the list of properties here when we validate referenced services:

https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractComponentNode.java#L679

Reply:
[https://issues.apache.org/jira/browse/NIFI-6798](url)